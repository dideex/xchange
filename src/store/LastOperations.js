import {observable} from 'mobx'
import openSocket from 'socket.io-client'

// last operations store
class LastOperations {
  @observable.ref data
  @observable loading
  
  constructor() {
    const url =
      process.env.NODE_ENV === 'development'
        ? 'http://localhost:3040'
        : 'http://176.119.158.145:3040'
    this.loading = true
    this.data = []
    // this.socket = openSocket('http://localhost:3040')
    this.socket = openSocket(url)
    this.socket.on('message', this._socketResolver)
  }

  _socketResolver = ({type, data, order}) => {
    switch (type) {
      case 'broadcast':
        this.loading = false
        return this.data = [...this.data, order]
      case 'init':
        this.loading = false
        return this.data = data
      default:
        return console.log('unknown message')
    }
  }
}

export const lastOperations = new LastOperations()

export default LastOperations
