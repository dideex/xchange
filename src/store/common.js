export const isNumber = num => {
  return /^\d+[\.]?\d*$/.test(num) || ''
}