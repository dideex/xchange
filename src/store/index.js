export {cash} from './Cash'
export {user} from './User'
export {lastOperations} from './LastOperations'
