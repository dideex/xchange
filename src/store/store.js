import {observable, action, computed} from 'mobx'

// menu state
class Store {
  @observable
  state // close | constructor | main | order

  constructor() {
    this.state = 'close'
  }
  // triggers opening or closing menu
  @action('toggle-state')
  toggleMenu = () => {
    this.state = this.state === 'close' ? 'main' : 'close'
  }

  // returns true when menu is opened
  @computed
  get isOpen() {
    return this.state !== 'close'
  }
}

const store = new Store()

export default store
export {Store}
