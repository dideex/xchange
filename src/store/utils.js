import Cookie from 'js-cookie'

export const setToken = (token, isAdmin) => {
  if (isAdmin) {
    Cookie.set('key', token)
    Cookie.set('isAdmin', true)
  } else Cookie.set('key', token)
}

export const logout = () => {
  Cookie.remove('key')
  Cookie.remove('isAdmin')
}

export const getToken = () => Cookie.get('key') || null
export const getAdminStatus = () => Cookie.get('isAdmin') || null
