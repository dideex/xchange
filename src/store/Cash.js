import {observable, action, computed} from 'mobx'
import openSocket from 'socket.io-client'
import Api from '../components/Api'
import {noty} from '../components/common'

// Mobx Cash store
// Keeps the current transittion payment data
// And currency rate exchange

class Cash {
  @observable inputValue
  @observable outputValue
  @observable currencyInput
  @observable currencyOutput
  @observable paymentStatus
  @observable draggedBadgeCurrency
  @observable orderId
  @observable loading
  @observable isNetworkError
  @observable.ref currency

  constructor() {
    this.inputValue = 0
    this.outputValue = 0
    this.outputValueInUsd = 0
    this.currencyInput = 0
    this.currencyOutput = 2
    this.paymentStatus = 0 // 0 - null, 1 - created, 2 - sended, 3 - closed
    this.draggedBadgeCurrency = null
    this.lessThenMinimal = false
    this.moreThenReseved = false
    this.orderId = null
    this.loading = false
    this.currency = []
    this.isNetworkError = false
    this.userRate = 0.9
    this._setInitalData()
  }

  _setInitalData = () => {
    const url =
      process.env.NODE_ENV === 'development'
        ? 'http://localhost:3040'
        : 'http://176.119.158.145:3040'
    this.socket = openSocket.connect(url)
    this.errorEmitter = Api.errorEmitter.bind(this)
    this.fetchCurrency()
  }

  // finds an equal value for input value
  // for 1BTC (with rate 0.9) should be 0.9BTC
  _calcOutput = value =>
    (value * this.currency[this.currencyInput].price_usd * this.userRate) /
    this.currency[this.currencyOutput].price_usd

  // finds an equal value for output value
  // for 1BTC (with rate 0.9) should be 1BTC
  // cur bitc 5448.77817486
  // should be to 0.9 (usd 6054.198)
  _calcInput = value =>
    (value * this.currency[this.currencyOutput].price_usd) /
    (this.currency[this.currencyInput].price_usd * this.userRate)

  // Swaps currencies values when changed currency is already set
  _reverseInputs = () => {
    const temp = this.outputValue / this.userRate
    this.outputValue = this.inputValue * this.userRate
    this.inputValue = temp
  }

  // finds an equal value for output value
  // for 1BTC (wtih rate 0.9) should be 0.9BTC
  _calcOutputWithoutRates = value =>
    (value * this.currency[this.currencyInput].price_usd * this.userRate) /
    this.currency[this.currencyOutput].price_usd

  // finds an equal value for input value
  // for 1BTC (wtih rate 0.9) should be 0.9BTC
  _calcInputWithoutRates = value =>
    (value * this.currency[this.currencyOutput].price_usd) /
    (this.currency[this.currencyInput].price_usd * this.userRate)

  // It holds total amount in common unit for userprofile
  _calcOutputInUsd = () =>
    (this.outputValueInUsd =
      this.outputValue * this.currency[this.currencyOutput].price_usd)

  // check for a correct number
  _isNumber = num => /^\d+[.]?\d{0,3}$/.test(num)

  // It returns true when input's currencies are equals
  _isSameCurrencyLabel = (id1, id2) =>
    this.currency[id1].label === this.currency[id2].label

  // Parses nubmer for "0,2" to "0.2", and " 0,2  " to "0.2"
  _parseNumber = num => {
    if (num === '') return 0
    else {
      num = num.replace(/\s/g, '')
      num = num.replace(',', '.')
    }
    return num
  }

  // It recalculates both input to correct range
  @action('Verify values posible limit')
  correctValuesLimits = () => {
    const {minimal} = this.currency[this.currencyInput]
    const {reserve} = this.currency[this.currencyOutput]
    if (this.inputValue < +minimal) {
      this.inputValue = minimal
      this.outputValue = this._calcOutput(minimal)
    }
    if (this.outputValue * this.userRate > +reserve) {
      this.outputValue = reserve
      this.inputValue = this._calcInput(reserve)
    }
  }

  @action('clear message error field')
  clearErr = () => (this.errorMessage = null)

  // Change and validate input, and recalculate output value
  @action('change input')
  changeInput = (number = 0) => {
    this.clearErr()
    const parsedNumber = this._parseNumber(number)
    if (this._isNumber(parsedNumber)) {
      this.inputValue = parsedNumber
      this.outputValue = this._calcOutput(parsedNumber)
    }
    this._calcOutputInUsd()
  }

  // Change and validate output, and recalculate input value
  @action('change output')
  changeOutput = (number = 0) => {
    this.clearErr()
    const parsedNumber = this._parseNumber(number)
    if (this._isNumber(parsedNumber)) {
      this.outputValue = parsedNumber
      this.inputValue = this._calcInput(parsedNumber)
    }
    this._calcOutputInUsd()
  }

  // Change output value, or switch it with input currency when are the same
  @action('set currency output')
  setCurrencyOutput = (id = 0) => {
    this.clearErr()
    if (this.currencyOutput === +id) return null
    if (this._isSameCurrencyLabel(id, this.currencyInput)) {
      this.currencyInput = this.currencyOutput
      this.currencyOutput = +id
      this._reverseInputs()
    } else {
      this.currencyOutput = +id
      this.outputValue = this._calcOutputWithoutRates(this.inputValue)
    }
  }

  // Change output value, or switch it with input currency when are the same
  @action('set currency input')
  setCurrencyInput = (id = 0) => {
    this.clearErr()
    if (this.currencyInput === +id) return null
    if (this._isSameCurrencyLabel(id, this.currencyOutput)) {
      this.currencyOutput = this.currencyInput
      this.currencyInput = +id
      this._reverseInputs()
    } else {
      this.currencyInput = +id
      this.inputValue = this._calcInputWithoutRates(this.outputValue)
    }
  }

  /**
   * Creates payment
   * Refetch currency rate exchanges
   * When token is not undefiended, creates protected order, otherwise creates public
   * @param {token: {String}, fromWallet: {String}, toWallet: {String}, email: {String}}{Object}
   * @return <Promise>
   * @public
   */
  @action('create payment and push to server')
  createPayment = ({token, fromWallet, toWallet, email}) => {
    this.clearErr()
    this.isNetworkError = false
    this.loading = true
    this.correctValuesLimits()
    this._calcOutputInUsd()
    return this.fetchCurrency().then(async () => {
      this.paymentStatus = 1
      const data = {
        inputValue: this.inputValue,
        outputValue: this.outputValue,
        outputValueInUsd: this.outputValueInUsd,
        currencyInput: this.currency[this.currencyInput].name,
        currencyOutput: this.currency[this.currencyOutput].name,
        currencyInputLabel: this.currency[this.currencyInput].label,
        currencyOutputLabel: this.currency[this.currencyOutput].label,
        paymentStatus: 1,
        email,
        fromWallet,
        toWallet,
        token,
      }
      if (token)
        await Api.post('orders', data, token)
          .then(
            this.errorEmitter(({result}) => {
              this.orderId = result._id
              noty('Ваш перевод успешно создан')
            }),
          )
          .catch(err => {
            console.error(err)
            this.isNetworkError = true
            noty('Ошибка создания', 'error')
          })
      else
        await Api.post('guestOrders', data)
          .then(
            this.errorEmitter(({result}) => {
              this.orderId = result._id
              noty('Ваш перевод успешно создан')
            }),
          )
          .catch(err => {
            console.error(err)
            this.isNetworkError = true
            noty('Ошибка создания', 'error')
          })
      this.loading = false
    })
  }

  /**
   * Fetch currency data and currency rate exchange from the server
   * Sorting result currencies
   * Replace id to number instead a name
   * @return <Promise>
   * @public
   */
  @action('get currency from the server')
  fetchCurrency = () =>
    new Promise((resolve, reject) => {
      this.loading = true
      return Api.get('currency')
        .then(
          this.errorEmitter(({data, userRate}) => {
            this.userRate = userRate
            this.currency = Object.values(data)
              .sort((a, b) => a.order - b.order)
              .map((row, i) => ({...row, id: i}))
            this.loading = false
            resolve()
          }),
        )
        .catch(err => {
          noty('Ошибка сервера', 'error')
          console.error(err)
          this.loading = false
          reject()
        })
    })

  /**
   * Creates boadcast query to socketio
   * @param {emailinputValue{String} outputValue{Sting|Number} inputLabel{String|Number} outputLabel{String} currency{String} paymentStatus{Number}}<data>
   * @public
   */
  @action('emit socket')
  emitSocket = data => this.socket.emit('newOrder', data)

  /**
   * Creates POST query to the server which means change a payment status for orderd to 2(order transmitted)
   * @param email{Sting}
   * @public
   */
  @action('confirm payment')
  cofirmPayment = email => {
    this.paymentStatus = 2
    const data = {
      _id: this.orderId,
      currency: this.currency[this.currencyOutput].icon,
      value: this.outputValue,
      email,
    }
    Api.post('confirmOrder', data)
      .then(
        this.errorEmitter(() => {
          noty('Ваш перевод передан на обработку')
          this.emitSocket({
            email,
            currency: this.currency[this.currencyOutput].icon,
            inputValue: this.inputValue,
            outputValue: this.outputValue,
            inputLabel: this.currency[this.currencyInput].label,
            outputLabel: this.currency[this.currencyOutput].label,
            paymentStatus: this.paymentStatus,
          })
        }),
      )
      .catch(err => {
        console.error(err)
        noty('Ошибка перевода', 'error')
      })
  }

  // helper for react-dnd, memoize the id which was dragged
  @action('drag badge')
  handleDragBadge = id => (this.draggedBadgeCurrency = id)

  // return formatted string with input value and their currency label is
  @computed
  get getInput() {
    return `${this.inputValue}, ${this.currency[this.currencyInput].label.toUpperCase()}`
  }
  // return formatted string with output value and their currency label is
  @computed
  get getOutput() {
    return `${this.outputValue}, ${this.currency[
      this.currencyOutput
    ].label.toUpperCase()}`
  }
  //returns formatted minimal amount for input value is
  @computed
  get getMinimalAmount() {
    const formatter = new Intl.NumberFormat('ru', 'currency')
    if (!this.currency.length) return null
    return `${formatter.format(this.currency[this.currencyInput].minimal)} ${
      this.currency[this.currencyInput].label
    }`
  }
  //returns formatted reserved amount for output value is
  @computed
  get getCurrencyReserve() {
    const formatter = new Intl.NumberFormat('ru', 'currency')
    if (!this.currency.length) return null
    return `${formatter.format(this.currency[this.currencyOutput].reserve)} ${
      this.currency[this.currencyOutput].label
    }`
  }
}

export const cash = new Cash()

export default Cash
