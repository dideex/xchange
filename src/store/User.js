import {observable, action} from 'mobx'
import Cookie from 'js-cookie'
import Api from '../components/Api'
import {noty} from '../components/common'

const setToken = (token, isAdmin) => {
  if (isAdmin) {
    Cookie.set('token', token)
    Cookie.set('isAdmin', true)
  } else Cookie.set('token', token)
}

const logout = () => {
  Cookie.remove('token')
  Cookie.remove('isAdmin')
}

const getToken = () => Cookie.get('token') || null
const getAdminStatus = () => Cookie.get('isAdmin') || null

// menu state
export default class User {
  @observable username
  @observable email
  @observable phone
  @observable wallets
  @observable lastOperations
  @observable convertedAmount
  @observable loading
  @observable token
  @observable isAdmin
  @observable errorMessage
  @observable.ref orders

  _setInitalData = () => {
    this.login = ''
    this.username = ''
    this.password = ''
    this.email = ''
    this.phone = ''
    this.wallets = {}
    this.lastOperations = []
    this.convertedAmount = 0
    this.loading = false
    this.token = ''
    this.isAdmin = false
    this.errorMessage = null
    this.orders = []
  }

  constructor() {
    this.token = getToken()
    this._setInitalData()
    this._checkToken()
    if (getAdminStatus()) this.isAdmin = true
    this.fetchData()
  }

  _checkToken = () => {
    const token = getToken()
    if (!token) return null
    Api.get('token', '', token)
      .then(response => response.json())
      .then(({success}) => {
        if (!success) {
          noty('Ошибка входа в аккаунт', 'error')
          logout()
        }
      })
      .catch(() => {
        noty('Ошибка сети', 'error')
        logout()
      })
  }

  @action('clear message error field')
  clearErr = () => (this.errorMessage = null)

  @action('change login')
  changeLogin = login => {
    this.clearErr()
    this.login = login
  }
  @action('change username')
  changeUsername = username => {
    this.clearErr()
    this.username = username
  }
  @action('change password')
  changePassword = password => {
    this.clearErr()
    this.password = password
  }
  @action('change email')
  changeEmail = email => {
    this.clearErr()
    this.email = email
  }
  @action('change phone')
  changePhone = phone => {
    this.clearErr()
    this.phone = phone
  }
  @action('change wallet')
  changeWallet = currencyLabel => value => {
    this.wallets[currencyLabel] = value
  }

  @action('sign out')
  signout = () => {
    this._setInitalData()
    noty('Вы вышли из аккаунта')
    logout()
  }

  @action('fetch user data from the server')
  fetchData = async () => {
    const token = getToken()
    if (!token) return null
    this.token = token

    await Api.get('userData', '', token)
      .then(res => res.json())
      .then(({wallets, lastOperations, username, email, login, convertedAmount}) => {
        this.login = login
        this.username = username
        this.email = email
        this.wallets = wallets || {}
        this.lastOperations = lastOperations || []
        this.convertedAmount = convertedAmount
      })
      .catch(() => noty('Ошибка сети', 'error'))
  }

  @action('udpate user data')
  updateInfo = () => {
    const token = getToken()
    if (!token) return null
    this.token = token

    const {username, wallets, email} = this
    return Api.post('userData', {username, wallets, email}, this.token).catch(() =>
      noty('Ошибка сети', 'error'),
    )
  }

  @action('Get token with username and passoword')
  getToken = () => {
    this.loading = true
    this.errorMessage = ''
    const {login: username, password} = this
    Api.post('signinUser', {username, password})
      .then(res => res.json())
      .then(({token, err, isAdmin}) => {
        this.loading = false
        if (token && !err) {
          this.token = token
          setToken(token, isAdmin)
          this.fetchData()
          if (isAdmin) {
            this.isAdmin = true
            noty('Hello admin!')
          } else {
            noty('Вы успешно вошли в аккаунт')
          }
        } else {
          noty(err, 'error')
        }
      })
      .catch(() => {
        this.loading = false
        noty('Ошибка сети', 'error')
      })
  }

  @action('Signup new user')
  signupUser = () => {
    const {login, email, username, password} = this
    Api.post('signupUser', {login, email, username, password})
      .then(res => res.json())
      .then(data => {
        const {token, err} = data
        if (!token) return noty(err, 'error')
        this.token = token
        noty('Вы успешно создали аккаунт')
        setToken(token)
        this.fetchData()
      })
      .catch(() => noty('Ошибка сети', 'error'))
  }

  @action('fetch orders by token')
  fetchOrdersByToken = async () => {
    if (!this.token) return null
    this.loading = true
    await Api.get('orders', '', this.token)
      .then(res => res.json())
      .then(data => {
        if (data.err) this.errorMessage = data.err
        else this.orders = data
      })
      .catch(err => {
        noty('Ошибка сети', 'error')
        console.error(err)
      })
    this.loading = false
  }

  @action('fetch order without token')
  fetchGuestOrder = async id => {
    this.loading = true
    const response = await Api.get('order', '', `?_id=${id}`)
      .then(res => res.json())
      .then(data => data)
      .catch(err => {
        noty('Ошибка сети', 'error')
        console.error(err)
      })
    this.loading = false
    if (response) return response
  }
}

export const user = new User()
