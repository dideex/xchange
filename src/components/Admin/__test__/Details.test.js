import React from 'react'
import Component from '../Details'
import {shallow} from 'enzyme'
// import {MobxProvider} from '../../../helpers/mobx'

import {cash} from '../../../store'

const fakeData = {
  id: 'Test id',
  inputValue: 'Test input',
  toWallet: 'Test wallet',
  fromWallet: 'Tests fromWallet',
  outputValue: 'Tests outputValue',
  email: 'Tests email',
  username: 'Tests username',
  paymentStatus: 0,
  loading: false,
  wallets: {},
  login: 'Test login',
  updatePaymentStatus() {},
}

describe('Admin details', () => {
  let cashStore = ''
  beforeAll(() => {
    cashStore = cash.create()
  })

  describe('Markup', () => {
    it('Base markup', () => {
      const wrapper = shallow(<Component cashStore={cashStore} date={{}} />)
      console.log(wrapper)
      // expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
