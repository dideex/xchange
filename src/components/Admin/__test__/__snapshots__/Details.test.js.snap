// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin details Markup Base markup 1`] = `"<div class=\\"css-stbtbn\\"><h3>Перевод № </h3><div class=\\"css-1n80iza\\"><div><div class=\\"css-1n80iza\\"><span>Кошелек с которого перевели:</span><strong></strong></div><div class=\\"css-1n80iza\\"><span>Сумма перевода:</span><strong>NaN </strong></div></div><div><div class=\\"css-1n80iza\\"><span>Получить на кошелек:</span><strong></strong></div><div class=\\"css-1n80iza\\"><span>Сумму получения:</span><strong>NaN </strong></div></div></div><div class=\\"css-x31cnd\\"><p><span>Login:</span><strong></strong></p><p><span>Email:</span><strong></strong></p><p><span>ФИО:</span><strong></strong></p><p><span>Статус:</span><strong></strong></p></div><div class=\\"css-luav6u\\"><span title=\\"Перевести в статус Ожидает перевода\\" color=\\"#FFE712\\" class=\\"css-9jaczz\\"></span><span title=\\"Перевести в статус Ожидает подтверждения\\" color=\\"#ff9800\\" class=\\"css-10hwe60\\"></span><span title=\\"Перевести в статус Переведено\\" color=\\"#8FBE00\\" class=\\"css-mlnvpn\\"></span><span title=\\"Перевести в статус Отменено\\" color=\\"#FC0000\\" class=\\"css-wv6hji\\"></span></div><ul class=\\"css-1fr0hla\\"></ul></div>"`;
